lib/pharos/cluster_manager.rb

```diff
@@ -85,29 +85,31 @@ def apply_phases
       # ca etc config files
       master_hosts = sorted_master_hosts
 
-      apply_phase(Phases::MigrateMaster, master_hosts, parallel: true)
-      apply_phase(Phases::ConfigureHost, config.hosts, master: master_hosts.first, parallel: true)
-      apply_phase(Phases::ConfigureClient, [master_hosts.first], master: master_hosts.first, parallel: false, optional: true)
+      apply_phase(Phases::MigrateMaster, master_hosts)
+      apply_phase(Phases::ConfigureHost, config.hosts, master: master_hosts.first)
+      apply_phase(Phases::ConfigureClient, [master_hosts.first], master: master_hosts.first, optional: true)
 
       unless @config.etcd&.endpoints
         # we need to use sorted etcd hosts because phases expects that first one has
         # ca etc config files
         etcd_hosts = sorted_etcd_hosts
-        apply_phase(Phases::ConfigureCfssl, etcd_hosts, parallel: true)
-        apply_phase(Phases::ConfigureEtcdCa, [etcd_hosts.first], parallel: false)
-        apply_phase(Phases::ConfigureEtcdChanges, [etcd_hosts.first], parallel: false)
-        apply_phase(Phases::ConfigureEtcd, etcd_hosts, parallel: true)
+        apply_phase(Phases::ConfigureCfssl, etcd_hosts)
+        apply_phase(Phases::ConfigureEtcdCa, [etcd_hosts.first])
+        apply_phase(Phases::ConfigureEtcdChanges, [etcd_hosts.first])
+        apply_phase(Phases::ConfigureEtcd, etcd_hosts)
       end
 
-      apply_phase(Phases::ConfigureSecretsEncryption, master_hosts, parallel: false)
-      apply_phase(Phases::SetupMaster, master_hosts, parallel: true)
-      apply_phase(Phases::UpgradeMaster, master_hosts, master: master_hosts.first, parallel: false) # requires optional early ConfigureClient
+      apply_phase(Phases::ValidateSecretsEncryption, master_hosts)
+      apply_phase(Phases::GenerateSecretsEncryptionKeys, ['localhost']) unless context['secrets_encryption']
```

Comment:
This runs only if `cluster_context` does not have a config. Generates the new config locally into cluster_context.